src/python/pants/backend/experimental/python/publish_pypi.py

```diff
@@ -0,0 +1,143 @@
+    register_interpreter_constraints = True
+    default_interpreter_constraints = ["CPython>=3.6"]
+
+
+class PypiRepositoryName(StringField):
+    alias = "repository"
+    help = "The pypi repository to upload the package to."
+    default = "pypi"
+
+
+class PypiRepositoryUrl(StringField):
+    alias = "repository_url"
+    help = "The pypi repository url to upload the package to. Takes priority over `repository`."
+
+
+class PypiRepositoryUsername(StringField):
+    alias = "username"
+    help = "The username for this repository. Suppirting ${BASH_STYLE} environment interpolation."
+    default = "${TWINE_USERNAME}"
+
+
+class PypiRepositoryPassword(StringField):
+    alias = "password"
+    help = "The password for this repository. Supporting ${BASH_STYLE} environment interpolation."
+    default = "${TWINE_PASSWORD}"
+
+
+class PypiPublishRequest(PublishRequest):
+    pass
+
+
+class PypiRepositoryTarget(PublishTarget):
+    alias = "pypi_repository"
+    help = "A pypi repository."
+
+    core_fields = (
+        *PublishTarget.core_fields,
+        PypiRepositoryName,
+        PypiRepositoryUrl,
+        PypiRepositoryUsername,
+        PypiRepositoryPassword,
+    )
+    publish_request_type = PypiPublishRequest
+
+
+@dataclass(frozen=True)
+class TwineCallArgs:
+    args: FrozenOrderedSet
+    env: FrozenDict[str, str]
+
+
+@rule
+async def twine_call_args(target: PypiRepositoryTarget) -> TwineCallArgs:
+    env = await Get(
+        Environment,
+        InterpolatedEnvironmentRequest(
+            {
+                "TWINE_USERNAME": target[PypiRepositoryUsername].value,
+                "TWINE_PASSWORD": target[PypiRepositoryPassword].value,
+            }
+        ),
+    )
+
+    args = ["--non-interactive"]
+
+    repo_name = target[PypiRepositoryName].value
+    repo_url = target[PypiRepositoryUrl].value
+    if repo_url:
+        args += ["--repository-url", repo_url]
+    else:
+        args += ["-r", repo_name]
+
+    return TwineCallArgs(FrozenOrderedSet(args), env)
+
+
+@rule
+async def publish_pypi(request: PypiPublishRequest, twine: Twine) -> PublishedPackage:
+    twine_pex = await Get(
+        VenvPex,
+        PexRequest(
+            output_filename="twine.pex",
+            internal_only=True,
+            requirements=PexRequirements(twine.all_requirements),
+            main=twine.main,
+        ),
+    )
+
+    call_args = await Get(TwineCallArgs, PypiRepositoryTarget, request.publish_target)
+    paths = [artifact.relpath for artifact in request.built_package.artifacts]
+
+    process = await Get(
+        Process,
+        VenvPexProcess,
+        VenvPexProcess(
+            twine_pex,
+            argv=["upload", *call_args.args, *paths],
+            input_digest=request.built_package.digest,
+            extra_env=call_args.env,
+            description=f"Publishing {', '.join(paths)} to {request.publish_target.address}.",
+        ),
+    )
+
+    await Get(ProcessResult, Process, process)
```

Comment:
This is side-effecting, which makes it a vaguely sketchy thing to do in a `@rule` rather than at the top level in the `@goal_rule` if there is any chance that the `Process` is not idempotent. For example: this process invoke will be cached (unless you pass a non-default `CacheScope` for the `*Process`), and will not re-run after succeeding. By default, failed processes are not cached, so that part should be fine. But it's also the case that while executing a cache lookup the `N+1th` time this process runs, we might retry the process and then cancel it.

So. An important question for this `Goal`'s API is: do you think that it is a reasonable requirement that all publishing commands/destinations MUST be idempotent? _If so_, then I think that the current API works. _If not so_, then rather than actually executing the publish processes in `@rule`s, you would want to return them as a list of publish `InteractiveProcess`es to have the `@goal_rule` execute synchronously in the foreground.

In my experience, idempotent publish commands are generally a good idea, but it's something to be cautious of here.

Reply:
Little unsure on how to proceed with this. 

Twine isn't idempotent, re-running the upload with the same inputs will result in an error.

So, I'm guessing I need to move the publish call from the `@rule` to the `@goal_rule`. Presumably by just returning a (boxed?) `InteractiveProcess` for the goal to execute.

But what about systems that do not necessarily use a process? i.e. the `docker` subsystem could in theory also use the docker library.

Reply:
I've moved the execution of the process to the `@goal_rule`.

Reply:
> Twine isn't idempotent, re-running the upload with the same inputs will result in an error.

True by default, although it can be configured to not fail for existing destinations. @benjyw , @jsirois : thoughts on this one?

> But what about systems that do not necessarily use a process? i.e. the docker subsystem could in theory also use the docker library.

I think that even if some of the implementations (like docker) are idempotent, it's still ok to execute them in the foreground. So this makes sense I think.

Reply:
Twine fails on re-upload for a reason, I'm not sure we want to elide that away unconditionally.